sunscreen/src/types/bfv/unsigned.rs

```diff
@@ -0,0 +1,358 @@
+};
+
+use crate as sunscreen;
+use crate::{
+    fhe::{with_fhe_ctx, FheContextOps},
+    types::{
+        ops::{
+            GraphCipherAdd, GraphCipherConstAdd, GraphCipherConstMul, GraphCipherConstSub,
+            GraphCipherMul, GraphCipherPlainAdd, GraphCipherPlainMul, GraphCipherPlainSub,
+            GraphCipherSub, GraphConstCipherSub, GraphPlainCipherSub,
+        },
+        Cipher,
+    },
+};
+use crate::{
+    types::{intern::FheProgramNode, BfvType, FheType, TypeNameInstance},
+    FheProgramInputTrait, Params, TypeName as DeriveTypeName, WithContext,
+};
+
+// TODO lil macro to gen all the ones from crypto_bigint?
+
+/// Unsigned 64 bit integer
+pub type Unsigned64 = Unsigned<1>;
+/// Unsigned 128 bit integer
+pub type Unsigned128 = Unsigned<2>;
+/// Unsigned 256 bit integer
+pub type Unsigned256 = Unsigned<4>;
+
+#[derive(Debug, Clone, Copy, DeriveTypeName, PartialEq, Eq)]
+/**
+ * A single unsigned integer.
+ */
+pub struct Unsigned<const LIMBS: usize> {
+    val: UInt<LIMBS>,
+}
+
+impl<const LIMBS: usize> NumCiphertexts for Unsigned<LIMBS> {
+    const NUM_CIPHERTEXTS: usize = 1;
+}
+
+impl<const LIMBS: usize> FheProgramInputTrait for Unsigned<LIMBS> {}
+impl<const LIMBS: usize> FheType for Unsigned<LIMBS> {}
+impl<const LIMBS: usize> BfvType for Unsigned<LIMBS> {}
+
+impl<const LIMBS: usize> std::fmt::Display for Unsigned<LIMBS> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.val)
+    }
+}
+
+impl<const LIMBS: usize> Default for Unsigned<LIMBS> {
+    fn default() -> Self {
+        Self::from(UInt::ZERO)
+    }
+}
+
+impl<const LIMBS: usize> TryIntoPlaintext for Unsigned<LIMBS> {
+    fn try_into_plaintext(
+        &self,
+        params: &Params,
+    ) -> std::result::Result<Plaintext, sunscreen_runtime::Error> {
+        let mut seal_plaintext = SealPlaintext::new()?;
+
+        let sig_bits = self.val.bits_vartime();
+        seal_plaintext.resize(sig_bits);
+
+        for i in 0..sig_bits {
+            let bit_value = self.val.bit_vartime(i);
+            seal_plaintext.set_coefficient(i, bit_value);
+        }
+
+        Ok(Plaintext {
+            data_type: self.type_name_instance(),
+            inner: InnerPlaintext::Seal(vec![WithContext {
+                params: params.clone(),
+                data: seal_plaintext,
+            }]),
+        })
+    }
+}
+
+impl<const LIMBS: usize> TryFromPlaintext for Unsigned<LIMBS> {
+    fn try_from_plaintext(
+        plaintext: &Plaintext,
+        params: &Params,
+    ) -> std::result::Result<Self, sunscreen_runtime::Error> {
+        let val = match &plaintext.inner {
+            InnerPlaintext::Seal(p) => {
+                if p.len() != 1 {
+                    return Err(sunscreen_runtime::Error::IncorrectCiphertextCount);
+                }
+
+                let bits = usize::min(std::mem::size_of::<UInt<LIMBS>>() * 8, p[0].len());
```

Comment:
plaintext values have 1024 to 32768 bits of precision depending on the scheme parameters, and taking the min here means you're effectively decoding values modulo 2^(64*LIMBS). I'm wondering if cases where the plaintext value is too large to fit in the specified limb it might be better to outright fail? We'll want to document this behavior in the user guide and eventually™️ make a cohesive and similar set of behaviors for the signed type (for which we'll need to make a signed `Int<LIMBS>` native type). Maybe that type will just wrap `UInt<Limbs>` and have a sign flag?

Reply:
I might be misunderstanding you, but I don't think we have the problem of plaintext values fitting into a particular limb. The output is a sum, coefficient by coefficient, of values like:
```
(UInt::from_u8(0x1) << i) * (UInt::from_u64(coeff))
```
so we don't build this up limb by limb (like the original version did when I initially opened the PR).

As I reread your comment, I'm thinking you mean "too large to fit into the specified _number of_ limbs". I.e. throw an error when `std::mem::size_of::<UInt<LIMBS>>() * 8 < p[0].len()`? If so, that does seem reasonable to me.